src/backend/distributed/metadata/metadata_utility.c

```diff
@@ -2357,3 +2365,1578 @@ IsForeignTable(Oid relationId)
+
+		/*
+		 * Only when the status of all depending jobs is done we clear this job and say
+		 * that is has no unmet dependencies.
+		 */
+		if (dependingJob->status == BACKGROUND_TASK_STATUS_DONE)
+		{
+			continue;
+		}
+
+		/*
+		 * we assume that when we ask for job to be cleared it has no dependencies that
+		 * have errorred. Once we move the job to error it should unschedule all dependant
+		 * jobs recursively.
+		 */
+		Assert(dependingJob->status != BACKGROUND_TASK_STATUS_ERROR);
+
+		hasUnmetDependency = true;
+		break;
+	}
+
+	systable_endscan(scanDescriptor);
+	table_close(pgDistBackgroundTasksDepend, AccessShareLock);
+
+	return hasUnmetDependency;
+}
+
+
+/*
+ * BackgroundTaskReadyToRun checks if a task is ready to run. This consists of two checks
+ *  - the task has no unmet dependencies
+ *  - the task either has no not_before value set, or the not_before time has passed.
+ *
+ * Due to the costs of checking we check them in reverse order, but conceptually they
+ * should be thought of in the above order.
+ */
+static bool
+BackgroundTaskReadyToRun(BackgroundTask *task)
+{
+	if (task->not_before)
+	{
+		if (*(task->not_before) > GetCurrentTimestamp())
+		{
+			/* task should not yet be run */
+			return false;
+		}
+	}
+
+	if (BackgroundTaskHasUmnetDependencies(task->jobid, task->taskid))
+	{
+		return false;
+	}
+
+	return true;
+}
+
+
+/*
+ * GetRunnableBackgroundTask returns the first candidate for a task to be run. When a task
+ * is returned it has been checked for all the preconditions to hold.
+ *
+ * That means, if there is no task returned the background worker should close and let the
+ * maintenance daemon start a new background tasks queue monitor once task become
+ * available.
+ */
+BackgroundTask *
+GetRunnableBackgroundTask(void)
+{
+	Relation pgDistBackgroundTasks =
+		table_open(DistBackgroundTaskRelationId(), AccessShareLock);
+
+	BackgroundTaskStatus taskStatus[] = {
+		BACKGROUND_TASK_STATUS_RUNNABLE
+	};
+
+	BackgroundTask *task = NULL;
+	for (int i = 0; !task && i < sizeof(taskStatus) / sizeof(taskStatus[0]); i++)
+	{
+		const int scanKeyCount = 1;
+		ScanKeyData scanKey[1] = { 0 };
+		const bool indexOK = true;
+
+		/* pg_dist_background_task.status == taskStatus[i] */
+		ScanKeyInit(&scanKey[0], Anum_pg_dist_background_task_status,
+					BTEqualStrategyNumber, F_OIDEQ, ObjectIdGetDatum(
+						BackgroundTaskStatusOid(taskStatus[i])));
+
+		SysScanDesc scanDescriptor =
+			systable_beginscan(pgDistBackgroundTasks,
+							   DistBackgroundTaskStatusTaskIdIndexId(),
+							   indexOK, NULL, scanKeyCount,
+							   scanKey);
+
+		HeapTuple taskTuple = NULL;
+		TupleDesc tupleDescriptor = RelationGetDescr(pgDistBackgroundTasks);
+		while (HeapTupleIsValid(taskTuple = systable_getnext(scanDescriptor)))
+		{
+			task = DeformBackgroundTaskHeapTuple(tupleDescriptor, taskTuple);
+			if (BackgroundTaskReadyToRun(task))
+			{
+				/* found task, close table and return */
+				break;
+			}
+			task = NULL;
```

Comment:
I was wondering whether we should free the task here, but I think it's fine because we're only reading tasks in runnable state anyway. There's should be very little of those.

Reply:
I had a pfree before, however simply pfree-ing a task might not be enough. There are some `char *` that this task might hold which don't free simply.

Besides only reading runnable tasks, it is also only called with memory contexts that clear timely.